[WGSL] Remove ResolveTypeReferences pass

tadeuzagallo opened the pull request:
#### 565fae35562a90c0c3956c51f21b3b02f38e8b63
<pre>
[WGSL] Remove ResolveTypeReferences pass
<a href="https://bugs.webkit.org/show_bug.cgi?id=253311">https://bugs.webkit.org/show_bug.cgi?id=253311</a>
rdar://106194074

Reviewed by Myles C. Maxfield.

Originally this pass was added to resolve which types referred to structs, which
was needed for the RewriteEntryPoints pass. Now that the type checker is a little
more capable we can use the types inferred from it instead and simplify the code.

* Source/WebGPU/WGSL/AST/ASTForward.h:
* Source/WebGPU/WGSL/AST/ASTNode.h:
* Source/WebGPU/WGSL/AST/ASTStringDumper.cpp:
* Source/WebGPU/WGSL/AST/ASTStringDumper.h:
* Source/WebGPU/WGSL/AST/ASTTypeName.h:
(WGSL::AST::NamedTypeName::name):
(isType):
(WGSL::AST::NamedTypeName::maybeResolvedReference const): Deleted.
(WGSL::AST::NamedTypeName::resolveTypeReference): Deleted.
* Source/WebGPU/WGSL/AST/ASTVisitor.cpp:
(WGSL::AST::Visitor::visit):
* Source/WebGPU/WGSL/AST/ASTVisitor.h:
* Source/WebGPU/WGSL/EntryPointRewriter.cpp:
(WGSL::EntryPointRewriter::rewrite):
(WGSL::EntryPointRewriter::collectStructs):
(WGSL::EntryPointRewriter::checkReturnType):
(WGSL::EntryPointRewriter::visit):
(WGSL::EntryPointRewriter::getResolvedType): Deleted.
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
* Source/WebGPU/WGSL/ResolveTypeReferences.cpp: Removed.
* Source/WebGPU/WGSL/ResolveTypeReferences.h: Removed.
* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::visit):
* Source/WebGPU/WGSL/WGSL.cpp:
(WGSL::prepareImpl):
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/261322@main">https://commits.webkit.org/261322@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/435bbbb3c3e02ea0e5f26b1a63f8d90f9719d352

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111280 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20422 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43825 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/2709 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120089 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21787 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11521 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/2297 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117045 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99361 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/103902 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/98136 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/31015 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/44754 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12922 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/32354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/86602 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13441 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/9344 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18878 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/51939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7843 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15396 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WGSL/ResolveTypeReferences.h:
```diff
@@ -1,34 +0,0 @@
-
-class ShaderModule;
-
-void resolveTypeReferences(ShaderModule&);
```
I'm kind of confused why it's OK to remove this. Any type can have a typedef that points to it, and the typedefs can chain. When typechecking, isn't it easier to not have to handle these chains of typedefs, but instead have them pre-resolved already?

tadeuzagallo replied:
You're right that we could keep this separate from type checking (although we'd need to move it before type checking). The reason I'm removing it is that I don't think it'd simplify the type checking to run this phase first, and the information that we'd need to add to the AST is less precise than the what type checking will add anyway.

tadeuzagallo committed:
[WGSL] Remove ResolveTypeReferences pass
https://bugs.webkit.org/show_bug.cgi?id=253311
rdar://106194074

Reviewed by Myles C. Maxfield.

Originally this pass was added to resolve which types referred to structs, which
was needed for the RewriteEntryPoints pass. Now that the type checker is a little
more capable we can use the types inferred from it instead and simplify the code.

* Source/WebGPU/WGSL/AST/ASTForward.h:
* Source/WebGPU/WGSL/AST/ASTNode.h:
* Source/WebGPU/WGSL/AST/ASTStringDumper.cpp:
* Source/WebGPU/WGSL/AST/ASTStringDumper.h:
* Source/WebGPU/WGSL/AST/ASTTypeName.h:
(WGSL::AST::NamedTypeName::name):
(isType):
(WGSL::AST::NamedTypeName::maybeResolvedReference const): Deleted.
(WGSL::AST::NamedTypeName::resolveTypeReference): Deleted.
* Source/WebGPU/WGSL/AST/ASTVisitor.cpp:
(WGSL::AST::Visitor::visit):
* Source/WebGPU/WGSL/AST/ASTVisitor.h:
* Source/WebGPU/WGSL/EntryPointRewriter.cpp:
(WGSL::EntryPointRewriter::rewrite):
(WGSL::EntryPointRewriter::collectStructs):
(WGSL::EntryPointRewriter::checkReturnType):
(WGSL::EntryPointRewriter::visit):
(WGSL::EntryPointRewriter::getResolvedType): Deleted.
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
* Source/WebGPU/WGSL/ResolveTypeReferences.cpp: Removed.
* Source/WebGPU/WGSL/ResolveTypeReferences.h: Removed.
* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::visit):
* Source/WebGPU/WGSL/WGSL.cpp:
(WGSL::prepareImpl):
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/261322@main
